### DISCUSSION TITLE
REST Data Source on Type Resolver

### DISCUSSION THREAD

**Discussion by aancer-rca** [apollographql/apollo-server#6015] [Q&A]
Hi,
I'm running into an error that reads as follows and was wondering if anyone here could offer some support: 
`[GraphQL error]: Message: Cannot read property 'userApi' of undefined, Location: [{"line":9,"column":5}], Path: actionAdded,user_tx`

I have my resolvers set up as such:
```
const resolvers = (pubsub: any): IResolvers => {
  return {
    JSON: JsonObject,
    Query: queries,
    Mutation: mutations(pubsub),
    Action: {  
      user_tx: async (action: IAction, args: any, { dataSources: { userApi } }: Context) => {
          return await userApi.getUser_tx(action.user_id);
        }
    },
    Subscription: subscriptions(pubsub)
  };
};
```

The Action resolver type here uses my dataSources and throws the error listed above. However, in my query resolvers the userApi works just fine. Am I missing something? Are REST Data Sources not made available to Field Resolvers that are returning from a subscription?

For example I looked at the types made available and they look like they provide a Context param type so I would think that perhaps the issue is that REST DataSources aren't instantiated prior to the Field resolvers since the following doesn't throw a syntax error:

```
const user_tx: IFieldResolver<IAction, Context, any> = async (action: IAction, args: any, { dataSources: { userApi } }: Context) => {
  return await userApi.getUser_tx(action.user_id);
};

const Action: IResolverObject = {
  user_tx
};

const resolvers = (pubsub: any): IResolvers => {
  return {
    JSON: JsonObject,
    Query: queries,
    Mutation: mutations(pubsub),
    Action,
    Subscription: subscriptions(pubsub)
  };
};
```

Perhaps the REST Data Source doesn't call the `Initialize` function prior to field resolvers that are returning from a subscription?

@glasser, @hwillson Do you either of you maybe know if REST Data Sources can be used with Field Resolvers that are returning from a subscription?

**Comment by aancer-rca** [comment]
Seems like this is a [known issue](https://github.com/apollographql/apollo-server/issues/1526) https://github.com/apollographql/apollo-server/issues/1526 that has yet to be resolved. Any traction on this?

**Comment by glasser** [comment]
Apollo Server does not currently support subscriptions. (Older versions supported them superficially; the superficial nature led to issues like this one.) There is documentation for using subscriptions-transport-ws in your server alongside Apollo Server (which soon will be rewritten to suggest the more maintained graphql-ws) but if that software doesn't support setting up data sources then it won't work. The data sources API is pretty simple so it should be straightforward to set up a REST data source in that software's equivalent of "context".

(We do hope to have integrated support for subscriptions in the future.)

**Reply by aancer-rca** [reply]
Thanks @glasser, I’ve added my data source to my context since and called ‘Initialize({})’ within and it’s working fine that way. Any chance you guys could add a warning of the limitation to the REST Data Source section of the docs?

**Reply by glasser** [reply]
I think we're working on a better solution where we help separate the REST Data Source a bit more from the rest of Apollo Server (which it shares no code with).
